app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/uploadsupportingdocs/UploadSupportingEvidenceController.scala

```diff
@@ -0,0 +1,1012 @@
+          )
+        )(
+          page = doYouWantToUploadSupportingEvidencePage(_, _)
+        )(
+          requiredPreviousAnswer               = { _ => Some(()) },
+          redirectToIfNoRequiredPreviousAnswer = controllers.returns.routes.TaskListController.taskList()
+        )
+      }
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+  def doYouWantToUploadSupportingDocumentsSubmit(): Action[AnyContent] = authenticatedActionWithSessionData.async {
+    implicit request =>
+      withUploadSupportingEvidenceAnswers(request) { (draftReturnId, _, _, fillingOutReturn, answers) =>
+        doYouWantToUploadSupportingDocumentsForm.bindFromRequest.fold(
+          errors =>
+            BadRequest(
+              doYouWantToUploadSupportingEvidencePage(errors, controllers.returns.routes.TaskListController.taskList())
+            ),
+          newDoYouWantToUploadSupportingEvidenceAnswer =>
+            if (newDoYouWantToUploadSupportingEvidenceAnswer) {
+              val updatedAnswers = answers match {
+                case IncompleteUploadSupportingEvidenceAnswers(None, supportingEvidences) =>
+                  IncompleteUploadSupportingEvidenceAnswers(
+                    Some(newDoYouWantToUploadSupportingEvidenceAnswer),
+                    supportingEvidences
+                  )
+                case IncompleteUploadSupportingEvidenceAnswers(
+                    Some(oldDoYouWantToUploadSupportingEvidenceAnswer),
+                    supportingEvidences
+                    ) =>
+                  if (newDoYouWantToUploadSupportingEvidenceAnswer =!= oldDoYouWantToUploadSupportingEvidenceAnswer) {
+                    IncompleteUploadSupportingEvidenceAnswers(
+                      Some(newDoYouWantToUploadSupportingEvidenceAnswer),
+                      List.empty
+                    )
+                  } else {
+                    IncompleteUploadSupportingEvidenceAnswers(
+                      Some(newDoYouWantToUploadSupportingEvidenceAnswer),
+                      supportingEvidences
+                    )
+                  }
+                case CompleteUploadSupportingEvidenceAnswers(
+                    oldDoYouWantToUploadSupportingEvidenceAnswer,
+                    supportingEvidences
+                    ) =>
+                  if (newDoYouWantToUploadSupportingEvidenceAnswer =!= oldDoYouWantToUploadSupportingEvidenceAnswer) {
+                    IncompleteUploadSupportingEvidenceAnswers(
+                      Some(newDoYouWantToUploadSupportingEvidenceAnswer),
+                      List.empty
+                    )
+                  } else {
+                    IncompleteUploadSupportingEvidenceAnswers(
+                      Some(newDoYouWantToUploadSupportingEvidenceAnswer),
+                      supportingEvidences
+                    )
+                  }
+              }
+
+              val newDraftReturn = fillingOutReturn.draftReturn match {
```

Comment:
you can save a lot of lines of code by just doing:
```scala
 fillingOutReturn.draftReturn.fold(
    _.copy(uploadSupportingDocuments = Some(updatedAnswers),
    _.copy(uploadSupportingDocuments = Some(updatedAnswers)
```
or if you really want to pattern match, just do:
```scala
  fillingOutReturn.draftReturn match {
    case s: DraftSingleDisposalReturn => ...
    case m: DraftMultipleDisposalReturn => ...
  }
```
This happens quite a bit in this controller

Reply:
Yes. I will looking something like this. Can be fixed afterwards.  Lot of copy and pasta in this controller. But the factoring out of common code will be done afterwards. I was busy trying to working how the UX should work as I was getting info that from the tickets.